File: simplestore/src/main/java/com/uber/simplestore/primitive/PrimitiveSimpleStore.java
/*
 * Copyright (C) 2019. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.simplestore.primitive;

import com.google.common.util.concurrent.ListenableFuture;
import com.uber.simplestore.SimpleStore;
import javax.annotation.CheckReturnValue;

/**
 * Store primitives on disk.
 *
 * <p>All methods never return null in the ListenableFuture, #{@link SimpleStore#contains(String)}
 * should be used for optionality. If the value is not set, the 0-byte primitive will be returned.
 */
public interface PrimitiveSimpleStore extends SimpleStore {

  @CheckReturnValue
  ListenableFuture<Integer> getInt(String key);

  @CheckReturnValue
  ListenableFuture<Integer> put(String key, int value);

  @CheckReturnValue
  ListenableFuture<Long> getLong(String key);

  @CheckReturnValue
  ListenableFuture<Long> put(String key, long value);

  @CheckReturnValue
  ListenableFuture<Boolean> getBoolean(String key);

  @CheckReturnValue
  ListenableFuture<Boolean> put(String key, boolean value);

  @CheckReturnValue
  ListenableFuture<Double> getDouble(String key);

  @CheckReturnValue
  ListenableFuture<Double> put(String key, double value);

  /**
   * Retrieves a #{@link java.nio.charset.StandardCharsets#UTF_16BE} string.
   *
   * @param key to fetch from
   * @return value if present, otherwise ""
   */
  @CheckReturnValue
  @Override
  ListenableFuture<String> getString(String key);

  /**
   * Store string as #{@link java.nio.charset.StandardCharsets#UTF_16BE}.
   *
   * <p>Putting "" will remove the value from disk.
   *
   * @param key name
   * @param value to store
   * @return stored value
   */
  @CheckReturnValue
  ListenableFuture<String> put(String key, String value);

  @CheckReturnValue
  @Override
  ListenableFuture<Void> remove(String key);
}

File: simplestore/src/main/java/com/uber/simplestore/primitive/PrimitiveSimpleStore.kt
/*
 * Copyright (C) 2019. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.simplestore.primitive

import com.google.common.util.concurrent.ListenableFuture
import com.uber.simplestore.SimpleStore

/**
 * Store primitives on disk.
 *
 * All methods never return null in the suspend functions, [SimpleStore.contains] should be used for optionality. 
 * If the value is not set, the 0-byte primitive will be returned.
 */
interface PrimitiveSimpleStore : SimpleStore {

    suspend fun getInt(key: String): Int

    suspend fun put(key: String, value: Int): Int

    suspend fun getLong(key: String): Long

    suspend fun put(key: String, value: Long): Long

    suspend fun getBoolean(key: String): Boolean

    suspend fun put(key: String, value: Boolean): Boolean

    suspend fun getDouble(key: String): Double

    suspend fun put(key: String, value: Double): Double

    /**
     * Retrieves a [java.nio.charset.StandardCharsets.UTF_16BE] string.
     *
     * @param key to fetch from
     * @return value if present, otherwise ""
     */
    override suspend fun getString(key: String): String

    /**
     * Store string as [java.nio.charset.StandardCharsets.UTF_16BE].
     *
     * Putting "" will remove the value from disk.
     *
     * @param key name
     * @param value to store
     * @return stored value
     */
    suspend fun put(key: String, value: String): String

    override suspend fun remove(key: String)
}

/**
 * Extension functions to provide ListenableFuture compatibility for Java callers
 */
fun PrimitiveSimpleStore.getIntFuture(key: String): ListenableFuture<Int> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { getInt(key) }
    )
}

fun PrimitiveSimpleStore.putIntFuture(key: String, value: Int): ListenableFuture<Int> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { put(key, value) }
    )
}

fun PrimitiveSimpleStore.getLongFuture(key: String): ListenableFuture<Long> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { getLong(key) }
    )
}

fun PrimitiveSimpleStore.putLongFuture(key: String, value: Long): ListenableFuture<Long> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { put(key, value) }
    )
}

fun PrimitiveSimpleStore.getBooleanFuture(key: String): ListenableFuture<Boolean> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { getBoolean(key) }
    )
}

fun PrimitiveSimpleStore.putBooleanFuture(key: String, value: Boolean): ListenableFuture<Boolean> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { put(key, value) }
    )
}

fun PrimitiveSimpleStore.getDoubleFuture(key: String): ListenableFuture<Double> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { getDouble(key) }
    )
}

fun PrimitiveSimpleStore.putDoubleFuture(key: String, value: Double): ListenableFuture<Double> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { put(key, value) }
    )
}

fun PrimitiveSimpleStore.putStringFuture(key: String, value: String): ListenableFuture<String> {
    return kotlinx.coroutines.guava.asListenableFuture(
        kotlinx.coroutines.GlobalScope.async { put(key, value) }
    )
}